### ISSUE TITLE
UnsupportedClassVersionError error - 52 vs 53 when run on Java 8

### ISSUE THREAD

**Issue by astubbs** [tlinkowski/UniJ#55]
When run on my CI setup, I get an error saying the UniJ library has been compiled with 53. When I check the class file locally with 0.1.1 I get 52 though.

Reference: https://github.com/confluentinc/parallel-consumer/runs/1724679799?check_suite_focus=true (link may have expired, but I can run it again)

```
13:15:11 19/01/21 in /Users/antony/repos/confluent/async-consumer-temp/async-consumer2
↪ javap -verbose .m2/repository/pl/tlinkowski/unij/pl.tlinkowski.unij.service.collect.jdk8/0.1.1/pl.tlinkowski.unij.service.collect.jdk8-0.1.1/pl/tlinkowski/unij/service/collect/jdk8/Jdk8UnmodifiableMapFactory.class | grep major
  major version: 52
```

```
2021-01-19T00:07:21.4717143Z 07:21.462  INFO  [main]  (ParallelEoSStreamProcessor.java:196#<init>) Confluent Parallel Consumer initialise... Options: ParallelConsumerOptions(consumer=io.confluent.csid.utils.LongPollingMockConsumer$MockitoMock$693145269@6eeeb9da, producer=org.apache.kafka.clients.producer.MockProducer$MockitoMock$899111225@17635531, ordering=UNORDERED, commitMode=PERIODIC_CONSUMER_SYNC, maxConcurrency=16, defaultMessageRetryDelay=PT1S)
2021-01-19T00:07:21.4963553Z [ERROR] Tests run: 43, Failures: 0, Errors: 34, Skipped: 9, Time elapsed: 1.654 s <<< FAILURE! - in io.confluent.parallelconsumer.ParallelEoSStreamProcessorTest
2021-01-19T00:07:21.4967558Z [ERROR] io.confluent.parallelconsumer.ParallelEoSStreamProcessorTest.produceMessageFlow(CommitMode)[1]  Time elapsed: 1.17 s  <<< ERROR!
2021-01-19T00:07:21.4971472Z java.lang.UnsupportedClassVersionError: pl/tlinkowski/unij/service/collect/jdk8/Jdk8UnmodifiableListFactory has been compiled by a more recent version of the Java Runtime (class file version 53.0), this version of the Java Runtime only recognizes class file versions up to 52.0
```

**Comment by tlinkowski** [comment]
I'm sorry, but I don't know why this happens. I'd need to investigate, but I'm afraid I won't find the time :/

My wild guess would be that the issue may be related to `module-info.class` being compiled to class version 53 (it's supposed to be the only file in version 53 in the library). I assumed JDK 8 would just ignore this file, but maybe it doesn't? 🤔

PS. You grepped `MapFactory` instead of `ListFactory`, but I assume it'd be the same for `ListFactory`.

**Comment by astubbs** [comment]
Pretty major issue if it doesn't work on java 8 ;)

Thinking out load here, I might try removing that file from the jar and see if I can get it to work. If it does, I'll look into it further, see if I can send a PR your way.

**Comment by tlinkowski** [comment]
Yes, it's a critical issue, I'm afraid :( But I really don't have the time to delve into this. However, if you manage to find out why exactly this happens, I should be able to release a fix.

PS. I checked the [files](https://search.maven.org/artifact/pl.tlinkowski.unij/pl.tlinkowski.unij.service.collect.jdk8/0.1.1/jar), and their versions seem totally OK:
![image](https://user-images.githubusercontent.com/8996855/105387650-0ce06300-5c16-11eb-990d-4c1b8a97ae72.png)

The only Java 9 file in the JAR is `module-info.class`:
![image](https://user-images.githubusercontent.com/8996855/105387783-2da8b880-5c16-11eb-9400-440d0c788408.png)

**Comment by astubbs** [comment]
Can that file be removed?

**Comment by tlinkowski** [comment]
I'm not sure what you mean:
- if you mean whether I can release the library without this file - yes, I can, but I'd like to be 100% sure it's the cause of this issue
- if you mean whether you can remove the file - you can just open the JAR in some archive manager (like 7-Zip) and remove the file from the JAR by pressing Delete

**Comment by kmcgovern-apixio** [comment]
I've seen a similar thing but i don't see the class file as version 52 when i run javap. i am using repo1.maven.org as the source for the jar.

```
$ curl -L https://repo1.maven.org/maven2/pl/tlinkowski/unij/pl.tlinkowski.unij.service.collect.jdk8/0.1.1/pl.tlinkowski.unij.service.collect.jdk8-0.1.1.jar | jar xv
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 35095  100 35095    0     0   183k      0 --:--:-- --:--:-- --:--:--  183k
  created: META-INF/
 inflated: META-INF/MANIFEST.MF
 inflated: META-INF/LICENSE
  created: META-INF/maven/
  created: META-INF/maven/pl.tlinkowski.unij/
  created: META-INF/maven/pl.tlinkowski.unij/pl.tlinkowski.unij.service.collect.jdk8/
 inflated: META-INF/maven/pl.tlinkowski.unij/pl.tlinkowski.unij.service.collect.jdk8/pom.properties
 inflated: META-INF/maven/pl.tlinkowski.unij/pl.tlinkowski.unij.service.collect.jdk8/pom.xml
  created: META-INF/services/
 inflated: META-INF/services/pl.tlinkowski.unij.service.api.collect.UnmodifiableListFactory
 inflated: META-INF/services/pl.tlinkowski.unij.service.api.collect.UnmodifiableMapFactory
 inflated: META-INF/services/pl.tlinkowski.unij.service.api.collect.UnmodifiableSetFactory
  created: pl/
  created: pl/tlinkowski/
  created: pl/tlinkowski/unij/
  created: pl/tlinkowski/unij/service/
  created: pl/tlinkowski/unij/service/collect/
  created: pl/tlinkowski/unij/service/collect/jdk8/
 inflated: pl/tlinkowski/unij/service/collect/jdk8/Jdk8UnmodifiableListFactory.class
 inflated: pl/tlinkowski/unij/service/collect/jdk8/Jdk8UnmodifiableMapFactory$Builder.class
 inflated: pl/tlinkowski/unij/service/collect/jdk8/Jdk8UnmodifiableMapFactory.class
 inflated: pl/tlinkowski/unij/service/collect/jdk8/Jdk8UnmodifiableSetFactory$Builder.class
 inflated: pl/tlinkowski/unij/service/collect/jdk8/Jdk8UnmodifiableSetFactory.class
 inflated: pl/tlinkowski/unij/service/collect/jdk8/package-info.class
 inflated: module-info.class
 inflated: pl/tlinkowski/unij/service/collect/jdk8/Jdk8UnmodifiableListFactory.class
 inflated: pl/tlinkowski/unij/service/collect/jdk8/Jdk8UnmodifiableMapFactory$Builder.class
 inflated: pl/tlinkowski/unij/service/collect/jdk8/Jdk8UnmodifiableMapFactory.class
 inflated: pl/tlinkowski/unij/service/collect/jdk8/Jdk8UnmodifiableSetFactory$Builder.class
 inflated: pl/tlinkowski/unij/service/collect/jdk8/Jdk8UnmodifiableSetFactory.class
$ find . -name '*.class' | xargs -t -n1 javap -verbose | grep -i major
javap -verbose ./pl/tlinkowski/unij/service/collect/jdk8/Jdk8UnmodifiableSetFactory.class
  major version: 53
javap -verbose ./pl/tlinkowski/unij/service/collect/jdk8/Jdk8UnmodifiableSetFactory$Builder.class
  major version: 53
javap -verbose ./pl/tlinkowski/unij/service/collect/jdk8/Jdk8UnmodifiableMapFactory.class
  major version: 53
javap -verbose ./pl/tlinkowski/unij/service/collect/jdk8/package-info.class
  major version: 52
javap -verbose ./pl/tlinkowski/unij/service/collect/jdk8/Jdk8UnmodifiableListFactory.class
  major version: 53
javap -verbose ./pl/tlinkowski/unij/service/collect/jdk8/Jdk8UnmodifiableMapFactory$Builder.class
  major version: 53
javap -verbose ./module-info.class
  major version: 53
  ```
  
  I can replicate the same problem locally and locally have fixed it with some hacks (removing module-info.java, setting compile target to 1.8). I haven't come up with a way to do this while preserving the existing java module functionality.

**Comment by tlinkowski** [comment]
Hey @kmcgovern-apixio , that's a wonderful discovery, thank you very much! In other words, the JAR I publish is simply broken. I'll take a look if it can be fixed at the stage of `compile*` tasks (I believe that - by definition - the other files just shouldn't be compiled to JDK9). However, if I can't fix it that way, I'll accept your fix of using `DuplicatesStrategy.EXCLUDE`. Thanks once again!

**Comment by tlinkowski** [comment]
Released to Maven Central as 0.1.3 (for some reason, 0.1.2 got skipped, but never mind).
